### PR TITLE
Improve DX and codebase quality

### PR DESCRIPTION
**Pull request by emagnier**
Here are a few things & improvements I had in a branch for a while. I'm opening a draft PR here to try to finalize that soon.

- [x] Add ESLint rules to improve codebase quality.
- [x] Provide project VSCode default settings & extension suggestions, to improve Developer eXperience (DX) on romcal (since VSCode is currently one of the most widely used free and open source IDE).
- [x] Fix/Improve TypeScript typing, and many ESLint warnings/errors in the codebase (fix https://github.com/romcal/romcal/issues/389).
- [x] ~~Add Jest snapshots, that test computed data on Liturgical/Gregorian calendars (over 3 consecutive years).~~ Already added here: https://github.com/romcal/romcal/pull/404
- [ ] ~~Improve Jest debugging with VSCode.~~
- [ ] ~~Include Prettier checks in CI (using GitHub Actions).~~

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
+module.exports = {
```
Comment (tukusejssirs) [inline .eslintrc.js:1]: Why did you rename `.eslintrc.json` to `.eslintrc.js`? I understand that both are supported, however, (currently) we don’t add any JS code, thus a plain JSON is sufficient. Of course, it requires to double-quote all keys and values.

File: lib/constants/colors.ts [hunk 2]
```diff
@@ -16,7 +16,7 @@ export enum Colors {
  */
 export const COLORS = Object.keys(Colors).filter(
   (key) => typeof Colors[key as keyof typeof Colors] === 'string',
-) as Array<Uppercase<keyof typeof Colors>>;
+) as Uppercase<keyof typeof Colors>[];
```
Comment (tukusejssirs) [inline lib/constants/colors.ts:19]: Why do you prefer using `Array<>` instead of `[]`? Is it an `eslint` setting?

File: lib/constants/ranks.ts [hunk 3]
```diff
@@ -57,9 +57,9 @@ export enum Ranks {
   Weekday = 'WEEKDAY',
 }
 
-export const RANKS = Object.keys(Ranks).filter((key) => typeof Ranks[key as keyof typeof Ranks] === 'string') as Array<
-  Uppercase<keyof typeof Ranks>
->;
+export const RANKS = Object.keys(Ranks).filter(
+  (key) => typeof Ranks[key as keyof typeof Ranks] === 'string',
+) as Uppercase<keyof typeof Ranks>[];
```
Comment (tukusejssirs) [inline lib/constants/ranks.ts:62]: You are inconsistent with using `Array<>` instead of `[]`. I prefer the latter, as it is shorter, however, it does not matter (AFAIK) which one we use, but I’d prefer using either one of them consistently.

File: lib/locales/index.ts [hunk 4]
```diff
@@ -15,7 +15,7 @@ import { locale as Sk } from './sk';
  * An BCP-47 IETF key to value mapping of all supported locale resource files
  * in romcal.
  */
-export const locales: Record<Lowercase<string>, Locale> = {
```
Comment (tukusejssirs) [inline lib/locales/index.ts:18]: Did you do make this change in order to fix #389? If yes, did you fix all the TS issues? If yes, consider adding `fix #389` to the PR description.

File: lib/models/calendar.ts [hunk 5]
```diff
@@ -134,10 +134,10 @@ export class Calendar implements BaseCalendar {
         // because of any general/particular calendar settings.
         // E.g. The 6th Thursday within the Easter Time can be not celebrated
         // because in some calendars, the Solemnity of the Ascension is taking precedence.
-        .reduce((acc, d) => {
+        .reduce<Date[]>((acc, d) => {
```
Comment (tukusejssirs) [inline lib/models/calendar.ts:137]: Do you want to use `Array<Date>` here to be consistent? :wink:

File: lib/models/calendar.ts [hunk 6]
```diff
@@ -166,7 +166,9 @@ export class Calendar implements BaseCalendar {
           // Retrieve calendar metadata from the proper of time
           const calendar: RomcalCalendarMetadata = isFromProperOfTime
             ? this.#buildCalendarMetadata(def, date, baseData)
-            : baseData!.calendar;
+            : // TODO: refactor this to avoir non-null assertion
```
Comment (tukusejssirs) [inline lib/models/calendar.ts:169]: Not important (as it should be later removed), but still:

```suggestion
            : // TODO: refactor this to avoid non-null assertion
```

File: lib/models/calendar.ts [hunk 7]
```diff
@@ -242,14 +244,14 @@ export class Calendar implements BaseCalendar {
     Object.keys(builtData.datesIndex).forEach((dateStr) => {
       // Order the LiturgicalDays objects, following the precedence rules defined in the UNLY #49.
       const dates: LiturgicalDay[] = builtData.datesIndex[dateStr]
-        .reduce((acc, key) => {
+        .reduce<LiturgicalDay[]>((acc, key) => {
```
Comment (tukusejssirs) [inline lib/models/calendar.ts:247]: Do you want to use `Array<LiturgicalDay>` here to be consistent? :wink:

File: lib/utils/string.ts [hunk 8]
```diff
@@ -29,21 +29,21 @@ export const toCamelCase = (string: string): string =>
...
-export const capitalize = (string: string): string => string.charAt(0).toUpperCase() + string.slice(1);
+export const toPascalCase = (string: string): string =>
+  capitalize(string.toLowerCase().replace(/((?:-|_|\s)\w)/g, (k: string) => k[1].toUpperCase()));
 
 /**
  * Converts the first character in the string, and the first character after un _, to a lowercase equivalent.
  * @param string
  */
-export const uncapitalize = (string: string): string =>
+export const unCapitalize = (string: string): string =>
```
Comment (tukusejssirs) [inline lib/utils/string.ts:48]: The _uncapitalise_ word does not sound good. Could we rename the function to something like `firstLetterToLowerCase()` or `firstToLowerCase()`?

File: lib/utils/string.ts [hunk 9]
```diff
@@ -29,21 +29,21 @@ export const toCamelCase = (string: string): string =>
   string.toLowerCase().replace(/((?:-|_|\s)\w)/g, (k: string) => k[1].toUpperCase());
 
 /**
- * Convert the string to PascalCase
+ * Converts the first character in the string to an uppercase equivalent.
  * @param string
  */
-export const toPascalCase = (string: string): string =>
-  capitalize(string.toLowerCase().replace(/((?:-|_|\s)\w)/g, (k: string) => k[1].toUpperCase()));
+export const capitalize = (string: string): string => string.charAt(0).toUpperCase() + string.slice(1);
```
Comment (tukusejssirs) [inline lib/utils/string.ts:35]: The _capitalise_ word denotes that the _entire_ string is changed to uppercase, however, that is not the case in here. Could we rename the function to something like `firstLetterToUpperCase()` or `firstToUpperCase()`?

File: tests/config.test.ts [hunk 10]
```diff
@@ -3,16 +3,16 @@ import Romcal from '../lib';
 const { RomcalConfig, LiturgicalDayConfig } = Romcal;
 
 describe('getConfig()', () => {
-  test('should get general config if country doesnt have default configurations', async () => {
+  test('should get general config if country does’t have default configurations', async () => {
```
Comment (tukusejssirs) [inline tests/config.test.ts]: I prefer using `'` in code comments, but not in localised strings (or any other string that is visible to the end user) I prefer using `’` (U+2019).

```suggestion
  test('should get general config if country doesn't have default configurations', async () => {
```

File: tests/dates.test.ts [hunk 11]
```diff
@@ -12,19 +14,28 @@ import Romcal from '../lib';
...
-      // Calculate the nb of days before the next target day (e.g. next Sunday after start)
-      - (7 + dayOfWeek - start.getUTCDay()) % 7,
-      0
-    ) / 7
+      Math.ceil((end.getTime() - start.getTime()) / 864e5) +
+        1 -
+        // Calculate the nb of days before the next target day (e.g. next Sunday after start)
+        ((7 + dayOfWeek - start.getUTCDay()) % 7),
+      0,
+    ) / 7,
```
Comment (tukusejssirs) [inline tests/dates.test.ts:36]: That lonely `1 -` on a line does not look nice IMHO.

**Review by tukusejssirs — CHANGES_REQUESTED** [review]
Great work, @emanier! :pray:

- Consider adding `fix #383` if you review that issue and the `eslint` config suggestions I provided there?
- Consider adding `fix #389` (TS 4.8 issues).
- While I don’t use VS Code (I prefer Sublime Text), however, I am okay with having `.vscode/` in romcal codebase.

File: lib/constants/colors.ts (hunk 2 above)
Comment (emagnier) [reply lib/constants/colors.ts:19]: This is the opposite. I prefer using `[]`. 

And yes, an `eslint` setting is enforcing it everywhere: `@typescript-eslint/array-type`.

File: lib/locales/index.ts (hunk 4 above)
Comment (emagnier) [reply lib/locales/index.ts:18]: I hadn't seen this issue. But yes indeed, this PR fixes all TS typing errors, and will close https://github.com/romcal/romcal/issues/389.

**Comment by emagnier** [comment]
Thanks for this first input. Please note that this is still a draft PR and things could move up a bit (so not fully ready for a review yet).

I will also take a look at your suggestions on #383.

File: lib/constants/colors.ts (hunk 2 above)
Comment (tukusejssirs) [reply lib/constants/colors.ts:19]: > This is the opposite. I prefer using `[]`.

Okay, then we should change it everywhere.

File: lib/locales/index.ts (hunk 4 above)
Comment (tukusejssirs) [reply lib/locales/index.ts:18]: I think you should consider watching _All Activity_ in this repository (as I do), so that you don’t miss a notification about any new issues/discussion/comment. I don’t always mention you. :wink:

**Comment by tukusejssirs** [comment]
> Please note that this is still a draft PR and things could move up a bit (so not fully ready for a review yet).

I understand that this PR is a WIP, however, I think a PR review might (or even should) be a looping process: the sooner a potential issue (or a suggestion, preference, etc) is noted, the better, as later (after a PR is finalised) it might be harder to fix that issue. I’ll try to find some time to review any commit you (or anyone else) makes in romcal as soon as I am notified. :wink:

File: lib/constants/colors.ts (hunk 2 above)
Comment (emagnier) [reply lib/constants/colors.ts:19]: This is actually the case.

Comment (tukusejssirs) [reply lib/constants/colors.ts:19]: Oh, you’re right. I have no idea why I thought it is not the case. Sorry for the noise then. :wink:

File: .eslintrc.js (hunk 1 above)
Comment (tukusejssirs) [inline .eslintrc.js:1]: You kept both `.eslintrc.js` and `.eslintrc.json` in the codebase. Did you just forget to delete `.eslintrc.json`?

File: lib/constants/colors.ts (hunk 2 above)
Comment (emagnier) [reply lib/constants/colors.ts:19]: No probs 😉

**Comment by tukusejssirs** [comment]
>     * [x]  Fix/Improve TypeScript typing, and many ESLint warnings/errors in the codebase.
> 
> fix #389

What about moving `fix #389` after the related change, for example:

> * [x]  Fix/Improve TypeScript typing, and many ESLint warnings/errors in the codebase (fix #389).

File: .eslintrc.js (hunk 1 above)
Comment (emagnier) [reply .eslintrc.js:1]: As I mentioned, this PR is still a work in progress (I didn't expect to get feedback so quickly). I only opened this PR to keep a visible track on it, but I haven't even taken the time myself to review and cleanup my work 😅 
However, thank you for your feedback. This is always very useful.

Comment (tukusejssirs) [reply .eslintrc.js:1]: > I didn't expect to get feedback so quickly

Well, I watch romcal repository (incl issues, discussions and anything in between) for I want to be notified of any and every change, and thus I can promptly (unless I am busy ATM) react to anything my opinion could be of some value.

> I only opened this PR to keep a visible track on it, but I haven't even taken the time myself to review and cleanup my work :sweat_smile:

Yes, that is my prefered workflow too: open a PR as early as possible in order to let others know of the state and progress of a particular work and to let others share there opinion and possibly suggest enhancements or provide better solutions. :wink:

> However, thank you for your feedback. This is always very useful.

Thanks! :pray:

File: lib/locales/index.ts (hunk 4 above)
Comment (emagnier) [reply lib/locales/index.ts:18]: This is already the case. But I have a very busy schedule, and I don't have time to go into all the subjects in depth and exchange on it (this is why I sometimes miss out certain points). 

At least, I try to skim though updates to keep an eye on the project. But I recognize that I rely a little on you 😄, because I know that you have more the issue tracker in mind than me.

File: .eslintrc.js [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+  extends: [
+    'eslint:recommended',
+    'plugin:@typescript-eslint/eslint-recommended',
+    'plugin:@typescript-eslint/recommended',
+    'plugin:jest/recommended',
+    'prettier',
+  ],
+  rules: {
+    'simple-import-sort/imports': 'error',
+    'simple-import-sort/exports': 'error',
```
Comment (tukusejssirs) [inline .eslintrc.js:23]: @emagnier, do you know any difference between [`eslint-plugin-simple-import-sort`](https://www.npmjs.com/package/eslint-plugin-simple-import-sort) and [`eslint-plugin-sort-keys-custom-order`](https://www.npmjs.com/package/eslint-plugin-sort-keys-custom-order)? Do you prefer the former?

In #383, I suggested to use the latter (I use it iny apps). I haven’t used `simple-import-sort` yet.

Note that I am not against using ``simple-import-sort`, I just want to know your opinion. :wink:

File: lib/utils/string.ts (hunk 8 above)
Comment (emagnier) [reply lib/utils/string.ts:48]: `.capitalize()` methods are well-known for this usage in programming languages, so I will keep this word (but rollback the change here to `uncapitalize`).

File: lib/utils/string.ts (hunk 9 above)
Comment (emagnier) [reply lib/utils/string.ts:35]: Same here. `.capitalize()` methods are well-known for this usage in programming languages, so I will keep this word.

File: tests/config.test.ts (hunk 10 above)
Comment (emagnier) [reply tests/config.test.ts]: I prefer to keep it that way here, so I don't have to use <code>`</code> or <code>"</code> here around the text description of this test.

File: tests/dates.test.ts (hunk 11 above)
Comment (emagnier) [reply tests/dates.test.ts:36]: Prettier rules. It doesn't bother me more than that, especially since it's happening only here.

File: .eslintrc.js (hunk 12 above)
Comment (emagnier) [reply .eslintrc.js:23]: I use `eslint-plugin-simple-import-sort` on several projects I'm working on. I like it for its simplified and opinionated configuration (like Prettier), and that a large community is also using it, that make things consistent between several projects.

I never used `eslint-plugin-sort-keys-custom-order`, but indeed I tend to prefer the former, which provide a more complete configuration by default without having to tweak it. Even if we can surely do the same with the second, with some configuration tweaks. But generally, more tweaks = more things to maintain on a long-term.

File: .eslintrc.js (hunk 1 above)
Comment (emagnier) [reply .eslintrc.js:1]: Because it's easier to read & edit JS than JSON. 

But if JS bother you, what about using YAML instead, like GitHub Actions?

Comment (emagnier) [reply .eslintrc.js:1]: FYI `jest.config.js` is also a JS file.

File: lib/utils/string.ts (hunk 9 above)
Comment (tukusejssirs) [reply lib/utils/string.ts:35]: Well, I still find it ambiguous, as _capitalisation_ means to _capitalise all letters in one or more words_, while what this function do is actually to keep the casing of all letters but the first letter of the first word is changed to upper-case, which is usually called _sentence case_.

[Dictionary.com](https://www.dictionary.com/browse/capitalise) says that _capitalise_ means _to write or print in capital letters letters or with an initial capital letter_, which confirms that your prefered sense is valid (I didn’t say it is not), however, it also confirms that its usage is ambiguous (it means either capitalisation of _all letters_ or _the initial/first letter_).

Comment (emagnier) [reply lib/utils/string.ts:35]: Yes, but it remains a common practice. So I'm going to keep what developers are used to seeing and using  :man_shrugging:.

File: .eslintrc.js (hunk 1 above)
Comment (tukusejssirs) [reply .eslintrc.js:1]: It does not bother me to use JS instead of JSON, however, I usually prefer using JSON instead of JS when I have no need to run some code in it (which I usually don’t).

As for YAML, I still prefer JSON. :wink:

As for readability, for me it is same if I read a JS object or a JSON, as long as I read them _both_ with nice syntax highlighting. As for editting, it is the same to me, as long as I have configured my text editor to use JSON schemas (for JSONs), I don’t need to really worry about double quotes (the editor auto-inserts them when I select a key from auto-suggestions even when I type the key name without quotes).

For for `jest.config.js` being a JS file, yes, it is, however, you didn’t modified it in the PR. However, I’d still prefer using JSON here too, unless we need to run some JS code in there.

Nevertheless, we can keep these files in JS. :wink:

File: tests/config.test.ts (hunk 10 above)
Comment (tukusejssirs) [reply tests/config.test.ts]: > I prefer to keep it that way here, so I don't have to use `` ` `` or `"` here around the text description of this test.

Pretty lazy you are! :rofl: I’d have no problem using either quote type, although I’d prefer usig `` ` ``.

That said, we can keep the typographic apostrophe, however, please do fix the typo. :wink:

```suggestion
  test('should get general config if country doesn’t have default configurations', async () => {
```

File: .eslintrc.js (hunk 12 above)
Comment (tukusejssirs) [reply .eslintrc.js:23]: Okay, I have no problem using either of them. :wink:

Just a note: I’ve just noticed I linked wrong plugin: `eslint-plugin-sort-keys-custom-order` sorts keys in _objects_ and _types_, not _imports. I Using [`eslint-plugin-import`](https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/order.md) to sort imports and `eslint-plugin-sort-keys-custom-order` to sort keys/properties in objects and types. :man_facepalming:

> But generally, more tweaks = more things to maintain on a long-term.

How so? I’ve configured it once and it is auto-fixable, thus no need to do anything on my side (just run `npm run lint`). Do you mean updating the package? A simple update script could do it (I don’t think an ESLint plugin code make my code non-executable). I am curious if you meant something else. :wink:

File: tests/config.test.ts (hunk 10 above)
Comment (emagnier) [reply tests/config.test.ts]: It was just to keep quoting type consistent here. But instead, I've another solution (or rather a workaround) :grinning::
```suggestion
  test('should get general config if country do not have default configurations', async () => {
```

Comment (tukusejssirs) [reply tests/config.test.ts]: > It was just to keep quoting type consistent here.

Well, I personally don’t care about quoting type consistence. I prefer using single quote, then template string, then double quotes, however, when I need/want to using one of them in a string, I have no problem to use either of them. :man_shrugging:

> But instead, I've another solution (or rather a workaround) :grinning:

Go for it! :wink:

File: .eslintrc.js (hunk 12 above)
Comment (emagnier) [reply .eslintrc.js:23]: It's a general fact. Keeping things simple make them always easiest to maintain. 

For example, when having less configuration code, we do not have to worry about updating this configuration if the plugin evolves (or is improved with new features), or if fashion coding style changes (on that point we lean on a community consensus and experience, which saves us from worrying about how to best configure things: the work is already done).

**Comment by emagnier** [comment]
Hello @tukusejssirs! If you agree, I would like to move forward on this PR, and merge it.

I'm running out of time currently to work on the last 2 todo items (listed on the top message of this PR). But I'll look into it as soon as possible in a future PR.

**Comment by tukusejssirs** [comment]
> Hello @tukusejssirs! If you agree, I would like to move forward on this PR, and merge it.

Okay. I’ve just gone through the changes again an it is good to go

> I'm running out of time currently to work on the last 2 todo items (listed on the top message of this PR). But I'll look into it as soon as possible in a future PR.

No problem with that. Just a question: you have checked `Add Jest snapshots, that test computed data on Liturgical/Gregorian calendars (over 3 consecutive years)`, but now it is striked through. Does it mean it is implemented in this PR or not?

**Review by tukusejssirs — APPROVED** [review]

**Comment by emagnier** [comment]
The Jest snapshots had already been added here https://github.com/romcal/romcal/pull/404 in a separated PR. Indeed, I've updated my top comment.
